### ISSUE TITLE
35191 - Trace

### ISSUE THREAD

**Issue by teamchong** [type-challenges/type-challenges#35247]
```ts
type Trace<T extends any[][]> = {[P in keyof T]: T[P][P & keyof T[P]]}[number]
```

[Playground](https://www.typescriptlang.org/play/?#code/PQKgUABBDMCsCMBOeEC0EAqAnAhgYwFNI1VSziAjATwgFcA7AEwIDMBLegxiACgAEGzdp0YBKCAGIAtlza0pYYhOUQAirQIBnAC5sA9vUVQMACwIRtuQhD0sIOCJoCOtHFnNScltgA8IbTQszR3kbO21gggAbAhl6bUCDfwSITw4IRjYcAHMDHCiAOmIACT0AdwIANwIsABpkgHJAzJZ2PFoo7Qs9CDx89qivcwjzTVCytgiLKgAHUaodWKLjHs8Aa2GTDmzAzTYpGZj6mO0miHdtWix6IPNo2IJ4xJuR1Jx0zJy8qIgJqYZ9C9ZgQikYIAAxPRYCAEHw4A4xABcYIABmiEsRtMCIABBLDQgC8EAA2sQoMT4LUAEwAXVqZJJ0FqABYacQ2VAsXNMFoukTsPgCAAePFYAB8EGAwBhPjmeG0XG6EAo5hQAB8IMziGiUWCJQA1NgEMo2G4AcUmxVoFEREBM2m0M00iKlCTwJgKACtNAUodlgHAkPAwCBgIpQBAAPrRmOxmMQACaeiuEAAwnpmBBijVzHG89GIKHFFzzALCEKMDKFUxAjh6FRiTTGxKiQBvYkABX8Nw2VFsmBptownabXYAZBBe-3hx2aTSAL7E+jyFVYNlgEv2CD8qzC4kU+q0+rEpmaudi8MgKP5uM8nRpnCaLTXm-xov7GZQrqb1sQACiLj5PUf6ygQ8oQPOEAsFgehSBADR8CWqDuvkMT0NkWjALQuhRJoDTFtifRPoERKklAIFytoQoAa4UQVruQr7pSEBHoy9SsjSYr1OqmpilxxAUWBVE0fk9GCoxxIAAzcdxdIklS9TSSxcknvUp6Seeil8fS5GgfK1GAXRZZ7sSDQ4A09QNBQFnwQ0KkNHgNkNE5dnHg0jBOQQTksHZnGWeZEAas5gXwT52lgOuEavrGEJXCM0IAMoKk6L7RYWYagMQEoJSYbjmH2KaaHoUTYYCzp2g6TousAboet6vpYP6gbIMAdaaBUWBZRAhrGo4xWlQY5X2o6zqupo7pej6foBggLVFSVuiDV1ACyULmKmuVRGhGFDZVo01eNdVTY1IZhkAA)

**Comment by StevenJPx2** [comment]
Damn that's really smart! Making the assumption of a square matrix removes any need for conditional typing!

**Comment by alexandroppolus** [comment]
A simple method to traverse an array/tuple in TS (`{[K in keyof T]...}`) is something that many (including me) often forget about

**Comment by teamchong** [comment]
Yes, in most scenarios, this is indeed the ideal approach for handling tuples, which have a maximum length of 9999. Recursive type declarations struggle with longer tuples as they quickly hit the recursion limit of TypeScript. I’ve restructured all my solutions to use the object construction method, which overcomes this limitation. Here are a few examples.

- #11875 
- #33163 
- #32794
